examples/solidity/ERC20/erc20.qnt

```diff
@@ -181,21 +196,163 @@ module erc20 {
+        state.allowance.keys().forall(a => isUint(state.allowance.get(a))),
+    }
 }
 
-// executing ERC20 transactions together with the mempool
-module mempool {
+// A set of tests to check the contract
+module erc20Tests {
     // We are using this module to test the contract,
     // so we fix the set of addresses right away.
-    pure val ADDR = Set("alice", "bob", "eve", "")
+    pure val ADDR = Set("0", "alice", "bob", "eve")
+    // import all definitions from erc20 for the fixed addresses
+    import erc20(AllAddresses = ADDR).*
+
     // We also restrict the amounts to restrict the search scope.
+    // When #566 is fixed, we would be able to use 256-bit integers.
     pure val AMOUNTS = 0.to(100)
+
+    var erc20State: erc20State
+
+    action init = all {
+        nondet sender = ADDR.exclude(Set(ZERO_ADDRESS)).oneOf()
+        nondet initialSupply = oneOf(AMOUNTS)
+        erc20State' = newErc20(sender, initialSupply),
+    }
+
+    // a helper action that assigns to state, provided that there are no errors
+    action fromResult(result: Erc20Result): bool = all {
+        result.returnedTrue,
+        result.error == "",
+        erc20State' = result,
+    }
+
+    action step = {
+        // run the contract methods
+        nondet sender = oneOf(ADDR)
+        nondet amount = oneOf(AMOUNTS)
+        any {
+            // transfer
+            nondet toAddr = oneOf(ADDR)
+            fromResult(erc20State.transfer(sender, toAddr, amount)),
+            // approve
+            nondet spender = oneOf(ADDR)
+            fromResult(erc20State.approve(sender, spender, amount)),
+            // transferFrom
+            nondet fromAddr = oneOf(ADDR)
+            nondet toAddr = oneOf(ADDR)
+            fromResult(erc20State.transferFrom(sender, fromAddr, toAddr, amount)),
+        }
+    }
+
+    // check the following invariants via:
+    // quint run --verbosity=3 --main=erc20Tests --invariant=noOverflowsInv erc20.qnt
+    val totalSupplyInv = isTotalSupplyCorrect(erc20State)
+
+    val zeroAddressInv = isZeroAddressEmpty(erc20State)
+
+    val noOverflowsInv = isNoOverflows(erc20State)
+
+    // Randomized tests that test whether postconditions holds true.
+
+    // Since we want to test ERC20 methods in arbitrary states,
+    // not just in the initial states, we have to describe "reasonable" states.
+    action initArbitrary = {
+        nondet owner = ADDR.exclude(Set(ZERO_ADDRESS)).oneOf()
+        nondet balances = ADDR.setOfMaps(AMOUNTS).oneOf()
+        nondet allowances = tuples(ADDR, ADDR).setOfMaps(AMOUNTS).oneOf()
+        erc20State' = {
+            balanceOf: balances,
+            totalSupply: sumOverBalances(balances),
+            allowance: allowances,
+            owner: owner,
+        }
+    }
+
+    // We want the contract to be only in "valid" states.
+    // So we have to define what it means.
+    pure def isValid(s: Erc20State): bool = and {
```

Comment:
I have found a way to define inductive invariants without saying "inductive invariants" 🎉